File: src/main/java/de/daniel/bactromod/config/ConfigScreen.java
package de.daniel.bactromod.config;

import de.daniel.bactromod.config.optiontypes.BooleanOption;
import de.daniel.bactromod.config.optiontypes.IntegerOption;
import net.minecraft.client.Minecraft;
import net.minecraft.client.OptionInstance;
import net.minecraft.client.Options;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.components.OptionsList;
import net.minecraft.client.gui.components.StringWidget;
import net.minecraft.client.gui.components.Tooltip;
import net.minecraft.client.gui.layouts.LinearLayout;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.gui.screens.options.OptionsSubScreen;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.Items;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public final class ConfigScreen {
    private ConfigScreen() {}

    public static OptionsSubScreen getConfigScreen(Screen parentScreen) {
        ConfigData data = Config.get();
        OptionInstance<?>[] configOptions = Arrays.stream(ConfigData.class.getDeclaredFields())
                .map(field -> configOption(field, data))
                .filter(java.util.Objects::nonNull)
                .toArray(OptionInstance<?>[]::new);

        return new OptionsSubScreen(
                parentScreen,
                Minecraft.getInstance().options,
                text("title")
        ) {
            @Override
            protected void addOptions() {
                if (this.list == null) return;
                this.list.addSmall(configOptions);

                this.list.addSmall(List.of(
                        Button.builder(
                                text("itemScalingFactors"),
                                _ -> Minecraft.getInstance().gui.setScreen(new ItemScalingScreen(this,
                                        Minecraft.getInstance().options, text("title").append(" - ")
                                        .append(text("itemScalingFactors"))))
                        ).tooltip(Tooltip.create(text("itemScalingFactors.desc"))).build()
                ));
            }
        };
    }

    private static OptionInstance<?> configOption(Field field, ConfigData data) {
        if (field.isAnnotationPresent(BooleanOption.class)) return booleanOption(field, data);
        IntegerOption option = field.getAnnotation(IntegerOption.class);
        return option == null ? null : integerOption(field, data, option);
    }

    private static OptionInstance<Boolean> booleanOption(Field field, ConfigData data) {
        String name = field.getName();
        try {
            return OptionInstance.createBoolean(key(name),
                    OptionInstance.cachedConstantTooltip(text(name + ".desc")), field.getBoolean(data),
                    value -> updateBoolean(field, data, value));
        } catch (IllegalAccessException e) {
            throw new IllegalStateException("Cannot access config field " + name, e);
        }
    }

    private static OptionInstance<Integer> integerOption(Field field, ConfigData data, IntegerOption option) {
        String name = field.getName();
        try {
            return new OptionInstance<>(key(name),
                    OptionInstance.cachedConstantTooltip(text(name + ".desc")), Options::genericValueLabel,
                    new OptionInstance.IntRange(option.min(), option.max()), field.getInt(data),
                    value -> updateInteger(field, data, value));
        } catch (IllegalAccessException e) {
            throw new IllegalStateException("Cannot access config field " + name, e);
        }
    }

    private static OptionInstance<Integer> itemScalingOption(String descriptionId, int value, ConfigData data) {
        return new OptionInstance<>(descriptionId, OptionInstance.cachedConstantTooltip(Component.translatable(
                "bactromod.options.itemScalingFactors.itemDesc", Component.translatable(descriptionId))),
                Options::genericValueLabel, new OptionInstance.IntRange(1, 100), value, newValue -> {
            data.itemScalingFactors.put(descriptionId, newValue);
            Config.save();
        });
    }

    private static MutableComponent text(String name) {
        return Component.translatable(key(name));
    }

    private static String key(String name) { return "bactromod.options." + name; }

    private static void updateBoolean(Field field, ConfigData data, boolean value) {
        try {
            field.setBoolean(data, value);
            Config.save();
        } catch (IllegalAccessException e) {
            throw new IllegalStateException("Cannot access config field " + field.getName(), e);
        }
    }

    private static void updateInteger(Field field, ConfigData data, int value) {
        try {
            field.setInt(data, value);
            Config.save();
        } catch (IllegalAccessException e) {
            throw new IllegalStateException("Cannot access config field " + field.getName(), e);
        }
    }

    private static class ItemScalingScreen extends OptionsSubScreen {
        private final List<ScalableItem> allItems = new ArrayList<>();
        private String searchFilter = "";

        ItemScalingScreen(Screen parentScreen, Options options, Component title) {
            super(parentScreen, options, title);
            loadItems();
        }

        private void loadItems() {
            ConfigData data = Config.get();
            Set<String> seen = new HashSet<>();
            for (Item item : BuiltInRegistries.ITEM) {
                if (item == Items.AIR) continue;
                String descId = item.getDescriptionId();
                String path = BuiltInRegistries.ITEM.getKey(item).getPath();
                addItem(data, seen, descId, path, data.itemScalingFactors.getOrDefault(descId, 100));
            }
            data.itemScalingFactors.forEach((descId, scaling) -> addItem(data, seen, descId, descId, scaling));
            allItems.sort(Comparator.comparing(ScalableItem::localizedName, String.CASE_INSENSITIVE_ORDER));
        }

        private void addItem(ConfigData data, Set<String> seen, String descId, String path, int scaling) {
            if (!seen.add(descId)) return;
            OptionInstance<Integer> option = itemScalingOption(descId, scaling, data);
            String localizedName = Component.translatable(descId).getString();
            allItems.add(new ScalableItem(descId, localizedName, path, option));
        }

        @Override
        protected void addTitle() {
            this.layout.setHeaderHeight(48);
            LinearLayout header = LinearLayout.vertical().spacing(4);
            header.defaultCellSetting().alignHorizontallyCenter();
            header.addChild(new StringWidget(this.title, this.font));

            EditBox searchBox = new EditBox(this.font, 200, 20, text("search"));
            searchBox.setHint(text("search").setStyle(EditBox.SEARCH_HINT_STYLE));
            searchBox.setValue(this.searchFilter);
            searchBox.setResponder(query -> {
                this.searchFilter = query;
                filterOptions(query);
            });
            header.addChild(searchBox);

            this.layout.addToHeader(header);
        }

        @Override
        protected void addContents() {
            this.list = this.layout.addToContents(new ItemOptionsList(this.minecraft, this.width, this));
            this.addOptions();
        }

        @Override
        protected void addOptions() {
            filterOptions(this.searchFilter);
        }

        private void filterOptions(String query) {
            if (!(this.list instanceof ItemOptionsList itemList)) return;

            itemList.clear();
            String filter = query.trim().toLowerCase(Locale.ROOT);
            itemList.addSmall(this.allItems.stream()
                    .filter(item -> filter.isEmpty() || item.matches(filter))
                    .map(ScalableItem::option)
                    .toArray(OptionInstance<?>[]::new));
            itemList.setScrollAmount(0);
        }
    }

    private static class ItemOptionsList extends OptionsList {
        ItemOptionsList(Minecraft minecraft, int width, OptionsSubScreen screen) {
            super(minecraft, width, screen);
        }

        void clear() {
            this.clearEntries();
        }
    }

    private record ScalableItem(String descId, String localizedName, String path, OptionInstance<Integer> option) {
        boolean matches(String query) {
            return localizedName.toLowerCase(Locale.ROOT).contains(query)
                    || path.toLowerCase(Locale.ROOT).contains(query)
                    || descId.toLowerCase(Locale.ROOT).contains(query);
        }
    }

}

File: src/main/kotlin/de/daniel/bactromod/config/ConfigScreen.kt
package de.daniel.bactromod.config

import de.daniel.bactromod.utils.SystemInfo
import de.daniel.bactromod.windowborder.DwmApi
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.screen.option.GameOptionsScreen
import net.minecraft.client.option.GameOptions
import net.minecraft.client.option.SimpleOption
import net.minecraft.text.Text

class ConfigScreen(screen: Screen) : GameOptionsScreen(
    screen, MinecraftClient.getInstance().options, Text.translatable("bactromod.options.title")
) {

    override fun addOptions() {
        this.body?.addAll(
            SimpleOption(
                "bactromod.options.gamma",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.gamma.description")),
                GameOptions::getGenericValueText,
                SimpleOption.ValidatingIntSliderCallbacks(0, 15),
                Config.load().gammaMultiplier
            ) { Config.save(Config.load().copy(gammaMultiplier = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.pumpkinblur",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.pumpkinblur.description")),
                Config.load().pumpkinBlur
            ) { Config.save(Config.load().copy(pumpkinBlur = it)) },

            SimpleOption(
                "bactromod.options.fireoffset",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.fireoffset.description")),
                GameOptions::getGenericValueText,
                SimpleOption.ValidatingIntSliderCallbacks(-100, 100),
                Config.load().fireOffset
            ) { Config.save(Config.load().copy(fireOffset = it)) },

            SimpleOption(
                "bactromod.options.shieldoffset",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.shieldoffset.description")),
                GameOptions::getGenericValueText,
                SimpleOption.ValidatingIntSliderCallbacks(-100, 100),
                Config.load().shieldOffset
            ) { Config.save(Config.load().copy(shieldOffset = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.boatmap",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.boatmap.description")),
                Config.load().showMapWhileInBoat
            ) { Config.save(Config.load().copy(showMapWhileInBoat = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.lavafog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.lavafog.description")),
                Config.load().lavaFog
            ) { Config.save(Config.load().copy(lavaFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.snowfog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.snowfog.description")),
                Config.load().powderSnowFog
            ) { Config.save(Config.load().copy(powderSnowFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.blindnessfog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.blindnessfog.description")),
                Config.load().blindnessFog
            ) { Config.save(Config.load().copy(blindnessFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.darknessfog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.darknessfog.description")),
                Config.load().darknessFog
            ) { Config.save(Config.load().copy(darknessFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.waterfog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.waterfog.description")),
                Config.load().waterFog
            ) { Config.save(Config.load().copy(waterFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.thickfog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.thickfog.description")),
                Config.load().thickFog
            ) { Config.save(Config.load().copy(thickFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.terrainfog",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.terrainfog.description")),
                Config.load().terrainFog
            ) { Config.save(Config.load().copy(terrainFog = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.fixShieldRiptideTrident",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.fixShieldRiptideTrident.description")),
                Config.load().fixShieldRiptideTrident
            ) { Config.save(Config.load().copy(fixShieldRiptideTrident = it)) },

            SimpleOption.ofBoolean(
                "bactromod.options.windowborders",
                SimpleOption.constantTooltip(Text.translatable("bactromod.options.windowborders.description")),
                Config.load().darkWindowBorders
            ) {
                Config.save(Config.load().copy(darkWindowBorders = it))
                if (SystemInfo.isWindows11()) DwmApi.updateDwm(MinecraftClient.getInstance().window.handle)
            }
        )
    }

}